rigging/connection.py

```diff
@@ -1,79 +1,136 @@
         """
         self.name = rig_name
-        self.sock_path = os.path.join(RIG_SOCK_DIR, self.name)
-        if not os.path.exists(self.sock_path):
-            raise OSError(f"No socket found for rig {self.name}")
+
+        # TODO: Check/handle errors
+        self._bus = dbus.SessionBus()
+
         try:
-            # if the rig has died but left its socket behind, this will fail
-            Client(self.sock_path)
-        except Exception:
-            raise DeadRigError(rig_name)
+            self._rig = self._bus.get_object(
+                        f"com.redhat.Rig.{rig_name}", f"/RigControl")
+        except dbus.exceptions.DBusException as exc:
+            if exc.get_dbus_name() == "org.freedesktop.DBus.Error.ServiceUnknown":
+                raise DBusServiceDoesntExistError(rig_name)
+            raise exc
+
+
 
     def _communicate(self, command):
         """
         Send a rig an instruction and then return the result directly to the
-        calling command (after json loading) for further handling
+        calling command for further handling
 
         :param command: The command to have the rig perform
         :return: The result of the command
         """
-        with Client(self.sock_path) as client:
-            try:
-                client.send_bytes(json.dumps(command).encode())
-            except Exception as err:
-                raise SendError(self.name, err)
-
-            try:
-                return json.loads(client.recv_bytes(4096).decode())
-            except Exception as err:
-                raise ResponseError(self.name, err)
-
-    def _create_instruction(self, instruction):
-        """
-        Create an instruction to send via _communicate() that can be serialized
-        via json
+        method = getattr(self._rig, command.name)
 
-        :param instruction: What to instruct the rig to do
-        :return: A dict formatted in the way a rig expects socket communication
-                 to look like
-        """
-        return {
-            'command': instruction,
-            'rig_name': self.name,
-        }
+        try:
+            ret = method(dbus_interface="com.redhat.RigInterface")
+            return RigDBusMessage(**ret)
+
+        except dbus.exceptions.DBusException as exc:
+            if exc.get_dbus_name() == "org.freedesktop.DBus.Error.UnknownMethod":
+                raise DBusMethodDoesntExistError(f"{command.name}()")
+            raise exc
+
+        except Exception as exc:
+            print(f"Exception caught while calling {command} on {self.name}:{exc}")
+
+        return None
 
     def destroy_rig(self):
         """
         Instruct the rig to self-terminate without triggering any configured
         actions or generating an archive.
         """
-        return self._communicate(self._create_instruction('destroy'))
+
+        return self._communicate(RigDBusCommand("destroy"))
+
+
+class RigDBusListener(dbus.service.Object):
```

Comment:
Also, for maintenance sake, let's get a docstring defined for this that explains the configuration and how to define new CLI-accessible methods.

Reply:
Done in d2d228fe8a6235c995f6e7734f13c10d2f23c8fc